oeps/oep-0050-hooks-extension-framework.rst

```diff
@@ -0,0 +1,227 @@
+edx-platform core. This proposal aims to be a solution to organize the majority
+of the changes that the authors have encountered after years of work extending
+the core in a way that is more compatible with the community at large, it is
+easier to reuse and far easier to migrate to newer versions of the codebase.
+
+This proposal draws inspiration from the cumulative experience of eduNEXT as the
+lead author, other members of the community, and from one of the largest
+extendable platforms for the web today for many years already, WordPress.
+
+.. _summary of the current options: https://github.com/edx/edx-platform/blob/master/docs/guides/extension_points.rst
+
+
+Specification
+=============
+
+The core of this proposed framework is the definition of a stable list of places
+in the life cycle of the learners registration, login, enrollment, course usage,
+and certification. Conversely, the authors life cycle will have a list of places
+during the creation and publication of courses. These places, called triggers
+will be written to the edx-platform code.
+Now, anyone that wants to extend the functionality of the platform at one of
+those places, can define a function, called an action, to be called when the
+trigger is executed. The sum of a trigger and action will be referred to as a
+hook.
+
+Triggers, broadly fit into two categories:
+
+* Triggers that allow actions called at this time to affect the application
+  flow via their results.
+* Triggers that will not allow the application flow to be altered by the
+  actions called.
+
+Although the first document contains a list of the proposed hooks, we now
+proposed that the documentation for this is kept in-code at https://github.com/edx/edx-platform/tree/master/docs/guides.
+This location should contain a document that covers the design principles and
+details that did not land in the OEP as well as the final decisions made in the
+OEP. Also will index the list of hooks and their current status. It should not
+cover the history and all the previous states of the framework. It should
+reflect the latest way of thinking and examples of how to use the framework.
+
+Additionally, a directory in the same location will include individual files
+that cover each available trigger.
+
+The signature of the functions (input/output) of every trigger should be
+maintained in a stable way. The same applies for additions or removals from the
+list of public triggers, which should follow a path of deprecation first and
+preferably be given an optional hook where the same context information or code
+execution path can be accessed.
+
+The mechanism by which a trigger is connected to an action will be a
+configurable python wrapper also contained in the platform core and by design
+must consider that:
+
+* Most actions will be implemented in plugins.
+* The order of execution for multiple actions listening on a trigger must be
+  respected, and the results of previously executed actions should be made
```

Comment:
This should be true only for the cases in which the actions run in a sync way (filters). For the async actions this should not apply, right?

Reply:
I made it more clear that the order will be kept, but the results will be ignored. This way we will run the pipeline as a whole async, but still keep the order so that others systems can rely on this.

An example would be pipeline with two actions running API calls in an CRM system:

- first action creates the user in the CRM
- second action attaches a piece of info to the user.

These two can run async, but they still need run in this order. Naturally the alternative would be to make one action that does the two things at once, but the whole point of having a pipeline is to be able to divide some work into smaller chunks that complete work by being added together. This again maximizes the flexibility by costing a bit of extra configuration

Reply:
With the latest edit async is no longer supported by the framework, so I'll mark this as resolved.